File: ios/Source/APIs/WalletLinkConnection.swift
// Copyright (c) 2018-2019 Coinbase, Inc. <https://coinbase.com/>
// Licensed under the Apache License, version 2.0

import BigInt
import CBCore
import CBCrypto
import CBHTTP
import CBStore
import os.log
import RxSwift

private typealias JoinSessionEvent = (sessionId: String, joined: Bool)

class WalletLinkConnection {
    private let userId: String
    private let notificationUrl: URL
    private let url: URL
    private let socket: WalletLinkWebSocket
    private let isConnectedObservable: Observable<Bool>
    private let joinSessionEventsSubject = PublishSubject<JoinSessionEvent>()
    private let requestsSubject = PublishSubject<HostRequest>()
    private var disposeBag = DisposeBag()
    private var metadata: [ClientMetadataKey: String]
    private let linkRepository: LinkRepository

    /// Incoming host requests for action
    let requestsObservable: Observable<HostRequest>

    /// Constructor
    ///
    /// - Parameters:
    ///     - url: WalletLink server URL
    ///     - websocketUrl: WalletLink websocket endpoint
    ///     - userId: User ID to deliver push notifications to
    ///     - notificationUrl: Webhook URL used to push notifications to mobile client
    ///     - metadata: client metadata forwarded to host once link is established
    ///     - linkRepository: manages links and dapps
    required init(
        url: URL,
        userId: String,
        notificationUrl: URL,
        metadata: [ClientMetadataKey: String],
        linkRepository: LinkRepository
    ) {
        self.url = url
        self.notificationUrl = notificationUrl
        self.userId = userId
        self.metadata = metadata
        self.linkRepository = linkRepository

        socket = WalletLinkWebSocket(url: url.appendingPathComponent("rpc"))
        requestsObservable = requestsSubject.asObservable()
        isConnectedObservable = socket.connectionStateObservable.map { $0.isConnected }

        socket.incomingRequestsObservable
            .flatMap { [weak self] dto -> Single<HostRequest?> in
                guard let self = self else { return Single.just(nil) }
                return self.linkRepository.getHostRequest(using: dto, url: self.url)
            }
            .unwrap()
            .subscribe(onNext: { [weak self] in self?.requestsSubject.onNext($0) })
            .disposed(by: disposeBag)

        observeConnection()
    }

    /// Stop connection when WalletLink instance is deallocated
    deinit {
        disposeBag = DisposeBag()
        _ = stopConnection().subscribe()
    }

    /// Connect to WalletLink server using parameters extracted from QR code scan
    ///
    /// - Parameters:
    ///     - sessionId: WalletLink host generated session ID
    ///     - secret: WalletLink host/guest shared secret
    ///
    /// - Returns: A single wrapping `Void` if connection was successful. Otherwise, an exception is thrown
    func link(sessionId: String, secret: String) -> Single<Void> {
        if let session = linkRepository.getSession(id: sessionId, url: url), session.secret == secret {
            return .justVoid()
        }

        return isConnectedObservable
            .do(onSubscribe: {
                self.linkRepository.saveSession(url: self.url, sessionId: sessionId, secret: secret)
            })
            .filter { $0 }
            .takeSingle()
            .flatMap { _ in self.joinSessionEventsSubject.filter { $0.sessionId == sessionId }.takeSingle() }
            .map { guard $0.joined else { throw WalletLinkError.invalidSession } }
            .timeout(15, scheduler: ConcurrentDispatchQueueScheduler(qos: .userInitiated))
            .logError()
            .do(onError: { _ in self.linkRepository.delete(url: self.url, sessionId: sessionId) })
    }

    /// Set metadata in all active sessions. This metadata will be forwarded to all the hosts
    ///
    /// - Parameters:
    ///   - key: Metadata key
    ///   - value: Metadata value
    ///
    /// - Returns: A single wrapping `Void` if operation was successful. Otherwise, an exception is thrown
    func setMetadata(key: ClientMetadataKey, value: String) -> Single<Void> {
        metadata[key] = value

        return linkRepository.getSessions(for: url)
            .compactMap { session -> Single<Bool>? in
                if let encryptedValue = try? value.encryptUsingAES256GCM(secret: session.secret) {
                    return self.socket.setMetadata(key: key, value: encryptedValue, for: session.id).logError()
                }

                print("Unable to encrypt \(key):\(value)")
                return nil
            }
            .zip()
            .asVoid()
    }

    /// Send signature request approval to the requesting host
    ///
    /// - Parameters:
    ///     - requestId: WalletLink host generated request ID
    ///     - responseData: Response data
    ///
    /// - Returns: A single wrapping `Void` if operation was successful. Otherwise, an exception is thrown
    func approve(requestId: HostRequestId, responseData: Data) -> Single<Void> {
        guard let session = linkRepository.getSession(id: requestId.sessionId, url: url) else {
            return .error(WalletLinkError.sessionNotFound)
        }

        let markEventAsSeen = linkRepository.markAsSeen(requestId: requestId, url: url)

        switch requestId.method {
        case .requestEthereumAccounts:
            guard let address = String(data: responseData, encoding: .utf8) else {
                return Single.error(WalletLinkError.missingResponseData)
            }

            let response = Web3ResponseDTO<[String]>(
                id: requestId.id,
                method: requestId.method,
                result: [address.lowercased()]
            )

            return markEventAsSeen
                .flatMap { _ in self.submitWeb3Response(response, session: session) }
                .flatMap { _ in
                    let dapp = Dapp(url: requestId.dappURL, name: requestId.dappName, logoURL: requestId.dappImageURL)
                    return self.linkRepository.saveDapp(dapp)
                }

        case .signEthereumMessage, .signEthereumTransaction, .submitEthereumTransaction:
            let response = Web3ResponseDTO<String>(
                id: requestId.id,
                method: requestId.method,
                result: responseData.toPrefixedHexString()
            )

            return markEventAsSeen.flatMap { _ in self.submitWeb3Response(response, session: session) }
        case .requestCanceled:
            return Single.error(WalletLinkError.unsupportedRequestMethodApproval)
        }
    }

    /// Send signature request rejection to the requesting host
    ///
    /// - Parameters:
    ///     - requestId: WalletLink host generated request ID
    ///
    /// - Returns: A single wrapping `Void` if operation was successful. Otherwise, an exception is thrown
    func reject(requestId: HostRequestId) -> Single<Void> {
        guard let session = linkRepository.getSession(id: requestId.sessionId, url: url) else {
            return .error(WalletLinkError.sessionNotFound)
        }

        let response = Web3ResponseDTO<String>(
            id: requestId.id,
            method: requestId.method,
            errorMessage: "User rejected signature request"
        )

        return linkRepository.markAsSeen(requestId: requestId, url: url)
            .flatMap { _ in self.submitWeb3Response(response, session: session) }
    }

    // MARK: - Connection management

    private func startConnection() -> Single<Void> {
        return Internet.statusChanges
            .filter { $0.isOnline }
            .takeSingle()
            .flatMap { _ in self.socket.connect() }
            .logError()
    }

    private func stopConnection() -> Single<Void> {
        return socket.disconnect()
            .logError()
            .catchErrorJustReturn(())
    }

    // MARK: - Session management

    private func joinSessions(sessions: [Session]) -> Single<Void> {
        return sessions.map { self.joinSession($0).asVoid().catchErrorJustReturn(()) }
            .zip()
            .map { _ in self.fetchPendingRequests() }
    }

    private func joinSession(_ session: Session) -> Single<Bool> {
        let credentials = Credentials(sessionId: session.id, secret: session.secret)

        return socket.joinSession(using: credentials.password, for: session.id)
            .flatMap { success -> Single<Bool> in
                guard success else {
                    self.joinSessionEventsSubject.onNext(JoinSessionEvent(sessionId: session.id, joined: false))

                    return .just(false)
                }

                return self.setSessionConfig(session: session)
            }
            .map { success in
                if success {
                    print("[walletlink] successfully joined session \(session.id)")

                    self.joinSessionEventsSubject.onNext(JoinSessionEvent(sessionId: session.id, joined: true))

                    return true
                } else {
                    print("[walletlink] Invalid session \(session.id). Removing...")

                    self.linkRepository.delete(url: self.url, sessionId: session.id)
                    self.joinSessionEventsSubject.onNext(JoinSessionEvent(sessionId: session.id, joined: false))

                    return false
                }
            }
            .logError()
    }

    private func setSessionConfig(session: Session) -> Single<Bool> {
        var encryptedMetadata = [String: String]()
        for (key, value) in metadata {
            guard let encryptedValue = try? value.encryptUsingAES256GCM(secret: session.secret) else {
                return .error(WalletLinkError.unableToEncryptData)
            }

            encryptedMetadata[key.rawValue] = encryptedValue
        }

        return socket.setSessionConfig(
            webhookId: userId,
            webhookUrl: notificationUrl,
            metadata: encryptedMetadata,
            for: session.id
        )
    }

    private func fetchPendingRequests() {
        _ = linkRepository.getSessions(for: url)
            .map { session in self.linkRepository.getPendingRequests(session: session) }
            .zip()
            .map { requests in requests.flatMap { $0 } }
            .logError()
            .subscribe(onSuccess: { requests in requests.forEach { self.requestsSubject.onNext($0) } })
    }

    // MARK: Request Handlers

    private func submitWeb3Response<T: Codable>(_ response: Web3ResponseDTO<T>, session: Session) -> Single<Void> {
        guard
            let json = response.asJSONString,
            let encryptedString = try? json.encryptUsingAES256GCM(secret: session.secret)
        else { return .error(WalletLinkError.unableToEncryptData) }

        return isConnectedObservable
            .filter { $0 }
            .takeSingle()
            .flatMap { _ in self.socket.publishEvent(.web3Response, data: encryptedString, to: session.id) }
            .map { guard $0 else { throw WalletLinkError.unableToSendSignatureRequestConfirmation } }
    }

    // MARK: - Observer(s)

    private func observeConnection() {
        var joinedSessionIds = Set<String>()
        let sessionSerialScheduler = SerialDispatchQueueScheduler(internalSerialQueueName: "serialSession")
        let connSerialScheduler = SerialDispatchQueueScheduler(internalSerialQueueName: "serialConn")

        let sessionChangesObservable = linkRepository.observeSessions(for: url)
            .distinctUntilChanged()
            .observeOn(connSerialScheduler)
            .concatMap { [weak self] sessions -> Single<[Session]> in
                guard let self = self else { return .just(sessions) }

                // If credentials list is not empty, try connecting to WalletLink server
                if !sessions.isEmpty {
                    return self.startConnection().map { sessions }.catchErrorJustReturn(sessions)
                }

                // Otherwise, disconnect
                return self.stopConnection().map { sessions }.catchErrorJustReturn(sessions)
            }

        Observable.combineLatest(isConnectedObservable, sessionChangesObservable)
            .debounce(0.3, scheduler: sessionSerialScheduler)
            .observeOn(sessionSerialScheduler)
            .concatMap { [weak self] isConnected, sessions -> Observable<Void> in
                guard let self = self else { return .justVoid() }

                if !isConnected {
                    joinedSessionIds.removeAll()
                    return .justVoid()
                }

                let currentSessionIds = Set<String>(sessions.map { $0.id })

                // remove unlinked sessions
                joinedSessionIds = joinedSessionIds.filter { currentSessionIds.contains($0) }

                let newSessions = sessions.filter { !joinedSessionIds.contains($0.id) }
                newSessions.forEach { joinedSessionIds.insert($0.id) }

                return self.joinSessions(sessions: newSessions).asObservable()
            }
            .catchErrorJustReturn(())
            .subscribe()
            .disposed(by: disposeBag)
    }
}

File: android/walletlink/src/main/java/com/coinbase/walletlink/apis/WalletLinkConnection.kt
// Copyright (c) 2018-2019 Coinbase, Inc. <https://coinbase.com/>
// Licensed under the Apache License, version 2.0

package com.coinbase.walletlink.apis

import com.coinbase.wallet.core.extensions.Strings
import com.coinbase.wallet.core.extensions.asUnit
import com.coinbase.wallet.core.extensions.takeSingle
import com.coinbase.wallet.core.extensions.toPrefixedHexString
import com.coinbase.wallet.core.extensions.unwrap
import com.coinbase.wallet.core.extensions.zipOrEmpty
import com.coinbase.wallet.crypto.extensions.encryptUsingAES256GCM
import com.coinbase.wallet.http.models.Credentials
import com.coinbase.wallet.http.connectivity.Internet
import com.coinbase.wallet.http.extensions.appendingPathComponent
import com.coinbase.walletlink.dtos.Web3ResponseDTO
import com.coinbase.walletlink.dtos.asJsonString
import com.coinbase.walletlink.exceptions.WalletLinkException
import com.coinbase.walletlink.extensions.create
import com.coinbase.walletlink.extensions.destroySession
import com.coinbase.walletlink.extensions.logError
import com.coinbase.walletlink.models.ClientMetadataKey
import com.coinbase.walletlink.models.Dapp
import com.coinbase.walletlink.models.EventType
import com.coinbase.walletlink.models.HostRequest
import com.coinbase.walletlink.models.HostRequestId
import com.coinbase.walletlink.models.RequestMethod
import com.coinbase.walletlink.models.Session
import com.coinbase.walletlink.repositories.LinkRepository
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.Observables
import io.reactivex.rxkotlin.addTo
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import timber.log.Timber
import java.net.URL
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

private data class JoinSessionEvent(val sessionId: String, val joined: Boolean)

internal class WalletLinkConnection private constructor(
    private val userId: String,
    private val notificationUrl: URL,
    private val linkRepository: LinkRepository,
    private val metadata: ConcurrentHashMap<ClientMetadataKey, String>,
    // Websocket connection URL
    val url: URL
) {
    private val requestsSubject = PublishSubject.create<HostRequest>()
    private val joinSessionEventsSubject = PublishSubject.create<JoinSessionEvent>()
    private val socket = WalletLinkWebSocket(url.appendingPathComponent("rpc"))
    private val disposeBag = CompositeDisposable()
    private val isConnectedObservable: Observable<Boolean> = socket.connectionStateObservable.map { it.isConnected }

    val disconnectSessionObservable = socket.disconnectSessionObservable

    /**
     * Incoming WalletLink host requests
     */
    val requestsObservable: Observable<HostRequest> = requestsSubject.hide()

    init {
        socket.incomingRequestsObservable
            .flatMap { linkRepository.getHostRequest(it, url).toObservable() }
            .unwrap()
            .map { request ->
                val hostRequestId = request.hostRequestId
                val session = linkRepository.sessions
                    .firstOrNull { it.url == hostRequestId.url && it.id == hostRequestId.sessionId }

                if (session?.version == null || request is HostRequest.DappPermission) {
                    return@map request
                }

                // for WalletLink v > 1, grab dapp details from EIP1102 request
                linkRepository.saveSession(
                    url = session.url,
                    sessionId = session.id,
                    secret = session.secret,
                    version = session.version,
                    dappName = hostRequestId.dappName,
                    dappImageURL = hostRequestId.dappImageURL,
                    dappURL = hostRequestId.dappURL
                )

                request
            }
            .subscribe { requestsSubject.onNext(it) }
            .addTo(disposeBag)

        observeConnection()
    }

    constructor(
        url: URL,
        userId: String,
        notificationUrl: URL,
        linkRepository: LinkRepository,
        metadata: Map<ClientMetadataKey, String>
    ) : this(
        userId = userId,
        notificationUrl = notificationUrl,
        linkRepository = linkRepository,
        metadata = ConcurrentHashMap(metadata),
        url = url
    )

    /**
     * Stop connection
     */
    fun disconnect() {
        disposeBag.clear()
        stopConnection().subscribe()
    }

    /**
     * Connect to WalletLink server using parameters extracted from QR code scan
     *
     * @param sessionId WalletLink host generated session ID
     * @param secret WalletLink host/guest shared secret
     * @param version WalletLink server version
     *
     * @return A single wrapping `Unit` if connection was successful. Otherwise, an exception is thrown
     */
    fun link(sessionId: String, secret: String, version: String?): Single<Unit> {
        val session = linkRepository.getSession(id = sessionId, url = url)

        if (session != null && session.secret == secret) {
            return Single.just(Unit)
        }

        return isConnectedObservable
            .doOnSubscribe {
                linkRepository.saveSession(
                    url = url,
                    sessionId = sessionId,
                    secret = secret,
                    version = version,
                    dappName = null,
                    dappImageURL = null,
                    dappURL = null
                )
            }
            .filter { it }
            .takeSingle()
            .flatMap { joinSessionEventsSubject.filter { it.sessionId == sessionId }.takeSingle() }
            .map { if (!it.joined) throw WalletLinkException.InvalidSession }
            .timeout(15, TimeUnit.SECONDS)
            .logError()
            .doOnError { linkRepository.delete(url, sessionId) }
    }

    /**
     * Set metadata in all active sessions. This metadata will be forwarded to all the hosts
     *
     * @param key Metadata key
     * @param value Metadata value
     *
     * @return A single wrapping `Unit` if operation was successful. Otherwise, an exception is thrown
     */
    fun setMetadata(key: ClientMetadataKey, value: String): Single<Unit> {
        metadata[key] = value

        return linkRepository.getSessions(url)
            .mapNotNull { session ->
                try {
                    val encryptedValue = value.encryptUsingAES256GCM(secret = session.secret)
                    return@mapNotNull socket.setMetadata(key, encryptedValue, session.id).logError()
                } catch (err: WalletLinkException.UnableToEncryptData) {
                    return@mapNotNull null
                }
            }
            .zipOrEmpty()
            .asUnit()
    }

    /**
    * Destroy session
    *
    * @param sessionId Session ID scanned offline (QR code, NFC, etc)
    *
    * @return A [Single] wrapping [Boolean] to indicate operation was successful
    */
    fun destroySession(sessionId: String): Single<Boolean> = socket.setSessionConfig(
            webhookId = null,
            webhookUrl = null,
            metadata = mapOf(ClientMetadataKey.Destroyed.rawValue to Strings.destroySession),
            sessionId = sessionId
        )

    /**
     * Send signature request approval to the requesting host
     *
     * @param requestId WalletLink host generated request ID
     * @param responseData Response data
     *
     * @return A single wrapping `Void` if operation was successful. Otherwise, an exception is thrown
     */
    fun approve(requestId: HostRequestId, responseData: ByteArray): Single<Unit> {
        val session = linkRepository.getSession(id = requestId.sessionId, url = url)
            ?: return Single.error(WalletLinkException.SessionNotFound)

        val markEventAsSeen = linkRepository.markAsSeen(requestId, url)

        return when (requestId.method) {
            RequestMethod.RequestEthereumAccounts -> {
                val address = responseData.toString(Charsets.UTF_8)
                val response = Web3ResponseDTO(
                    id = requestId.id,
                    method = requestId.method,
                    result = listOf(address.toLowerCase())
                )

                return markEventAsSeen
                    .flatMap { submitWeb3Response(response, session) }
                    .flatMap {
                        val dapp = Dapp(requestId.dappURL, requestId.dappName, requestId.dappImageURL)
                        linkRepository.saveDapp(dapp)
                    }
            }
            RequestMethod.SignEthereumMessage,
            RequestMethod.SubmitEthereumTransaction,
            RequestMethod.SignEthereumTransaction -> {
                val response = Web3ResponseDTO(
                    id = requestId.id,
                    method = requestId.method,
                    result = responseData.toPrefixedHexString()
                )

                return markEventAsSeen.flatMap { submitWeb3Response(response, session) }
            }
            RequestMethod.RequestCanceled -> {
                Single.error(WalletLinkException.UnsupportedRequestMethodApproval)
            }
        }
    }

    /**
     * Send signature request rejection to the requesting host
     *
     * @param requestId WalletLink host generated request ID
     *
     * @return A single wrapping `Void` if operation was successful. Otherwise, an exception is thrown
     */
    fun reject(requestId: HostRequestId): Single<Unit> {
        val session = linkRepository.getSession(id = requestId.sessionId, url = url)
            ?: return Single.error(WalletLinkException.SessionNotFound)

        val response = Web3ResponseDTO<String>(
            id = requestId.id,
            method = requestId.method,
            errorMessage = "User rejected signature request"
        )

        return linkRepository.markAsSeen(requestId, url)
            .flatMap { submitWeb3Response(response, session = session) }
    }

    // Connection management

    private fun startConnection(): Single<Unit> {
        return Internet.statusChanges
            .filter { it.isOnline }
            .takeSingle()
            .flatMap { socket.connect() }
            .logError()
    }

    private fun stopConnection(): Single<Unit> {
        return socket.disconnect()
            .logError()
            .onErrorReturn { Single.just(Unit) }
    }

    // Session management

    private fun joinSessions(sessions: List<Session>): Single<Unit> = sessions
        .map { joinSession(it).asUnit().onErrorReturn { Unit } }
        .zipOrEmpty()
        .map { fetchPendingRequests() }

    private fun joinSession(session: Session): Single<Boolean> {
        val credentials = Credentials.create(session.id, session.secret)

        return socket.joinSession(credentials.password, session.id)
            .flatMap { success ->
                if (!success) {
                    joinSessionEventsSubject.onNext(JoinSessionEvent(sessionId = session.id, joined = false))
                    return@flatMap Single.just(false)
                }

                return@flatMap setSessionConfig(session = session)
            }
            .map { success ->
                if (success) {
                    Timber.i("[walletlink] successfully joined session ${session.id}")

                    joinSessionEventsSubject.onNext(JoinSessionEvent(sessionId = session.id, joined = true))

                    return@map true
                } else {
                    Timber.i("[walletlink] Invalid session ${session.id}. Removing...")

                    linkRepository.delete(url = url, sessionId = session.id)
                    joinSessionEventsSubject.onNext(JoinSessionEvent(sessionId = session.id, joined = false))

                    return@map false
                }
            }
            .logError()
    }

    private fun setSessionConfig(session: Session): Single<Boolean> {
        val encryptedMetadata = HashMap<String, String>()
        for ((key, value) in metadata) {
            try {
                encryptedMetadata[key.rawValue] = value.encryptUsingAES256GCM(secret = session.secret)
            } catch (err: Exception) {
                return Single.error(WalletLinkException.UnableToEncryptData)
            }
        }

        return socket.setSessionConfig(
            webhookId = userId,
            webhookUrl = notificationUrl,
            metadata = encryptedMetadata,
            sessionId = session.id
        )
    }

    private fun fetchPendingRequests() {
        linkRepository.getSessions(url)
            .map { linkRepository.getPendingRequests(it) }
            .zipOrEmpty()
            .map { requests -> requests.flatten() }
            .logError()
            .subscribeBy(onSuccess = { requests -> requests.forEach { requestsSubject.onNext(it) } })
            .addTo(disposeBag)
    }

    // Request Handlers

    private inline fun <reified T> submitWeb3Response(response: Web3ResponseDTO<T>, session: Session): Single<Unit> {
        val json = response.asJsonString()
        val encryptedString: String

        try {
            encryptedString = json.encryptUsingAES256GCM(secret = session.secret)
        } catch (exception: Exception) {
            return Single.error(WalletLinkException.UnableToEncryptData)
        }

        return isConnectedObservable
            .filter { it }
            .takeSingle()
            .flatMap { socket.publishEvent(EventType.Web3Response, encryptedString, session.id) }
            .map { if (!it) throw WalletLinkException.UnableToSendSignatureRequestConfirmation }
    }

    // Observers

    private fun observeConnection() {
        var joinedSessionIds = HashSet<String>()
        val sessionSerialScheduler = Schedulers.single()
        val connSerialScheduler = Schedulers.single()

        val sessionChangesObservable = linkRepository.observeSessions(url)
            .distinctUntilChanged()
            .observeOn(connSerialScheduler)
            .concatMap { sessions ->
                // If credentials list is not empty, try connecting to WalletLink server
                if (sessions.isNotEmpty()) {
                    return@concatMap startConnection().map { sessions }.onErrorReturn { sessions }.toObservable()
                }

                // Otherwise, disconnect
                return@concatMap stopConnection().map { sessions }.onErrorReturn { sessions }.toObservable()
            }

        Observables.combineLatest(isConnectedObservable, sessionChangesObservable)
            .observeOn(sessionSerialScheduler)
            .debounce(300, TimeUnit.MILLISECONDS)
            .concatMap { (isConnected, sessions) ->
                if (!isConnected) {
                    joinedSessionIds.clear()
                    return@concatMap Observable.just(Unit)
                }

                val currentSessionIds = HashSet(sessions.map { it.id })

                // remove unlinked sessions
                joinedSessionIds = joinedSessionIds.filterTo(HashSet()) { currentSessionIds.contains(it) }

                val newSessions = sessions.filter { !joinedSessionIds.contains(it.id) }
                newSessions.forEach { joinedSessionIds.add(it.id) }

                joinSessions(sessions = newSessions).toObservable()
            }
            .onErrorReturn { Unit }
            .subscribe()
            .addTo(disposeBag)
    }
}
